src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -225,26 +226,26 @@ class LibraryService (protected val argUserInfo: UserInfo,
     }
   }
 
-  def findDocuments(criteria: LibrarySearchParams): Future[PerRequestMessage] = {
-    getEffectiveDiscoverGroups(samDao) flatMap { userGroups =>
-      // we want docsFuture and ids to be parallelized - so declare them here, outside
-      // of the for-yield.
-      val docsFuture = searchDAO.findDocuments(criteria, userGroups)
-      val idsFuture = rawlsDAO.getWorkspaces
-
-      (for {
-        docs <- docsFuture
-        ids <- idsFuture
-      } yield RequestComplete(updateAccess(docs, ids))) recover {
-        case ex =>
-          throw new FireCloudExceptionWithErrorReport(ErrorReport(errorMessageFromSearchException(ex)))
-      }
+  def searchFor(criteria: LibrarySearchParams, searchMethod:(LibrarySearchParams, Seq[String], Map[String, UserPolicy])=>Future[LibrarySearchResponse]): Future[PerRequestMessage] ={
+    val workspacePolicies: Future[Seq[UserPolicy]] = samDao.listWorkspaceResources
+    workspacePolicies flatMap { policyList =>
+      val workspacePolicyMap = (policyList map { policy =>
+        (policy.resourceId.value, policy)
+      }).toMap
+      getEffectiveDiscoverGroups(samDao) map { userGroups =>
+        // we want docsFuture and ids to be parallelized - so declare them here, outside
+        // of the for-yield.
+        searchMethod(criteria, userGroups, workspacePolicyMap)
+      } map (RequestComplete(_))
```

Comment:
could you parallelize the two calls to Sam? `listWorkspaceResources`  and `getEffectiveDiscoverGroups` are independent of each other, but in this code are serialized.  Maybe something like:
```
  def searchFor(criteria: LibrarySearchParams, searchMethod:(LibrarySearchParams, Seq[String], Map[String, UserPolicy])=>Future[LibrarySearchResponse]): Future[PerRequestMessage] ={
    val workspacePoliciesFuture: Future[Map[String, UserPolicy]] = samDao.listWorkspaceResources map { policyList =>
      (policyList map { policy =>
        (policy.resourceId.value, policy)
      }).toMap
    }
    val userGroupsFuture: Future[Seq[String]] = getEffectiveDiscoverGroups(samDao)

    for {
      workspacePolicyMap <- workspacePoliciesFuture
      userGroups <- userGroupsFuture
      searchResults <- searchMethod(criteria, userGroups, workspacePolicyMap)
    } yield {
      RequestComplete(searchResults)
    }
  }
```